geth/jail/jail.go

```diff
@@ -143,9 +157,13 @@ func (jail *Jail) Call(chatID, this, args string) string {
 }
 
 // Send is a wrapper for executing RPC calls from within Otto VM.
+// It uses own jail's VM instance instead of cell's one to
+// increase safety of cell's vm usage.
+// TODO(divan): investigate if it's possible to do conversions
+// withouth involving otto code at all.
 // nolint: errcheck, unparam
-func (jail *Jail) Send(call otto.FunctionCall, vm *vm.VM) otto.Value {
-	request, err := vm.Call("JSON.stringify", nil, call.Argument(0))
+func (jail *Jail) Send(call otto.FunctionCall) otto.Value {
+	request, err := jail.vm.Call("JSON.stringify", nil, call.Argument(0))
```

Comment:
Even if it works, it's pretty strange. You have an object `call.Argument(0)` from one vm and you call `JSON.stringify` with this object in completely different VM. I think it opens a way to some kind of bugs like missing references etc.

Reply:
@adambabik yeah, I originally thought this as well, and tracked down the code executed there. In this case there is no access to `runtime`/otto's vm is involved, so it's safe. I agree that it's completely unclear from looking this code, but that's exactly the reason why I want to move all these JSON-RPC conversions into separate package, so it could be properly isolated and documented.
It's safe so far, but we need more clarity on that in case this piece of code will evolve.

Reply:
> It's safe so far, but we need more clarity on that in case this piece of code will evolve.

So far :) I would prefer to solve it properly than delaying the problem. 

I think we can safely assume that `jail.Send` is never called directly except for `sendAsync`. What we could do is to have two methods: `jail.Send(otto.FunctionCall, otto.VM)` and `jail.SendAsync(otto.FunctionCall, vm.VM)`. Notice `otto.VM`, it will be fine and properly guarded because this is called from already safe context. `jail.SendAsync` will be called from a goroutine, thus `vm.VM` is used to acquire a lock.

What do you think about this?